### ISSUE TITLE
Error: invalid character 'ï'

### ISSUE THREAD

**Issue by Daniel-CS-Team** [SpecterOps/BloodHound#30]
When ingesting data from our domain collected with the latest Sharphound (v2.0.0) the following error is logged in BloodhoundCE:
`{"level":"error","time":"2023-08-15T07:39:55.366631981Z","message":"Error processing ingest task 178: invalid character 'ï' looking for beginning of value"}`

A quick grep for the character in question revealed several user displaynames containing this character. As far as I can tell, the error renders the entire ingest corrupt and makes it unusable.

Edit: Not only is the current ingest corrupt, but the entire neo4j-Database becomes unusabel for Bloodhound and must be rebuild from scratch.

As Sharphound collects the data with this encoding, I would expect Bloodhound to be able to parse it without problems.

**Comment by dmquik** [comment]
Was having the same issue, but only when uploading via the API. If I uploaded via the GUI then it was fine, but had a few files over 1gb so the API was the only option.

I ended up using the script written by aconite33 linked to in this issue: https://github.com/SpecterOps/BloodHound/issues/20
Then had to change lines 62 and 63 as below, and then the uploads via the API worked without that special character:

`with open(args.file, mode='r', encoding='utf-8-sig') as f:
        data = f.read().encode('utf-8')`

**Comment by Daniel-CS-Team** [comment]
Thank you @dmquik
I had to adapt your approach for large files, but it works now. For anyone looking for a solution:
```python
with open('newly_encoded.json', mode='w', encoding='utf-8') as target_file:
        with open('original.json', encoding='utf-8-sig') as source_file:
            while True:
                data = source_file.read(1024)
                if not data:
                    break
                target_file.write(data)
```

Anyhow, it remains a bug in my opinion.

**Comment by rvazarkar** [comment]
I'm not entirely sure whats going on with this because I wrote a test ensuring that umlauts ingest fine.

https://github.com/SpecterOps/BloodHound/pull/32

The test works a-ok and it uses the API to do the ingest. I'm wondering if theres something else going on here

**Comment by MrCarney** [comment]
In case someone else stumbles upon this like I did, the "ï" character that the error message was referring to in my instance was the [UTF-8 BOM](https://en.wikipedia.org/wiki/Byte_order_mark).

I was using PowerShell to read the JSON files to send them via the API. PowerShell 5.1 and lower interpret any UTF-8 encoding as UTF-8 BOM. I was able to get around it by reading the file as text then getting the bytes with the UTF8 encoder.

**Comment by Daniel-CS-Team** [comment]
I digged a bit deeper into the problem, because i did a rework of our ingestion code yesterday. 
Disclaimer: I am using Python for this, but it seems other languages have similar issues.

First of all, I do not think it is the BOM itself. I have 3 reasons for this assumption:
1. The character "ï" (0xC3AF in Unicode) is not part of any legit BOM.
2. Not all files nor domains trigger the error, so it must be specific to a domain. The encoding is the same in all files.
3. When trying to parse the files content to an UTF-8 string in Python, I encountered issues way into the file. The BOM is per definition the first bytes in a file.

Amazingly, the workaround from @dmquik  is still working. But it is also needed.

It appears, that files containing the Bytestring 0xC3AF cannot be parsed correctly by Bloodhound and are therefore discarded. That means, if you have a user in your domain with that character in her display name for example, you will not be able to ingest the file without further preprocessing.
